arbor/include/arbor/simd/sve.hpp

```diff
@@ -0,0 +1,900 @@
+         copy_to(y, b);
+
+        for (unsigned i = 0; i<len; ++i) {
+            r[i] = std::pow(a[i], b[i]);
+        }
+        return copy_from(r);
+    }
+
+    static unsigned simd_width(const svfloat64_t& m) {
+        return svlen_f64(m);
+    }
+
+protected:
+    // Compute n and f such that x = 2^n·f, with |f| ∈ [1,2), given x is finite and normal.
+    static svint32_t logb_normal(const svfloat64_t& x) {
+        svuint32_t xw    = svtrn2_u32(svreinterpret_u32_f64(x), svreinterpret_u32_f64(x));
+        svuint64_t lmask = svdup_n_u64(0x00000000ffffffff);
+        svuint64_t xt = svand_u64_z(svptrue_b64(), svreinterpret_u64_u32(xw), lmask);
+        svuint32_t xhi = svreinterpret_u32_u64(xt);
+        auto emask = svdup_n_u32(0x7ff00000);
+        auto ebiased = svlsr_n_u32_z(svptrue_b64(), svand_u32_z(svptrue_b64(), xhi, emask), 20);
+
+        return svsub_s32_z(svptrue_b64(), svreinterpret_s32_u32(ebiased), svdup_n_s32(1023));
+    }
+
+    static svfloat64_t fraction_normal(const svfloat64_t& x) {
+        svuint64_t emask = svdup_n_u64(-0x7ff0000000000001);
+        svuint64_t bias =  svdup_n_u64(0x3ff0000000000000);
+        return svreinterpret_f64_u64(
+            svorr_u64_z(svptrue_b64(), bias, svand_u64_z(svptrue_b64(), emask, svreinterpret_u64_f64(x))));
+    }
+
+    static inline svfloat64_t horner1(svfloat64_t x, double a0) {
+        return add(x, broadcast(a0));
+    }
+
+    static inline svfloat64_t horner(svfloat64_t x, double a0) {
+        return broadcast(a0);
+    }
+
+    template <typename... T>
+    static svfloat64_t horner(svfloat64_t x, double a0, T... tail) {
+        return fma(x, horner(x, tail...), broadcast(a0));
+    }
+
+    template <typename... T>
+    static svfloat64_t horner1(svfloat64_t x, double a0, T... tail) {
+        return fma(x, horner1(x, tail...), broadcast(a0));
+    }
+
+    static svfloat64_t fms(const svfloat64_t& a, const svfloat64_t& b, const svfloat64_t& c) {
+        return svnmsb_f64_z(svptrue_b64(), a, b, c);
+    }
+};
+
+}  // namespace detail
+
+namespace simd_abi {
+template <typename T, unsigned N> struct sve;
+template <> struct sve<double, 0> {using type = detail::sve_double;};
+template <> struct sve<int, 0> {using type = detail::sve_int;};
+
+template <typename T> struct reg_type;
+template <> struct reg_type<int> { using type = svint64_t;};
+template <> struct reg_type<double> { using type = svfloat64_t;};
+
+template <typename T> struct mask_type;
+template <> struct mask_type<int> { using type = svbool_t;};
+template <> struct mask_type<double> { using type = svbool_t;};
+
+template<typename Type> struct type_to_impl;
+template<> struct type_to_impl<svint64_t> { using type = detail::sve_int;};
+template<> struct type_to_impl<svfloat64_t> { using type = detail::sve_double;};
+template<> struct type_to_impl<svbool_t> { using type = detail::sve_mask;};
+};  // namespace simd_abi
+
+using namespace arb::simd::simd_abi;
+using arb::simd::detail::simd_traits;
```

Comment:
These `using` statements will lift things into `arb::simd` that don't really belong there. I'd recommend renaming `type_to_impl` as `sve_type_to_impl`, putting it in the `detail` namespace, and using it qualified below.

Reply:
Okay